File: app/src/main/java/com/relic/presentation/displaysubs/subslist/SubItemAdapter.java
package com.relic.presentation.displaysubs.subslist;

import android.view.ViewGroup;
import android.widget.ImageView;

import androidx.annotation.NonNull;
import androidx.databinding.BindingAdapter;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.RecyclerView;

import com.relic.domain.models.SubredditModel;
import com.relic.presentation.displaysubs.DisplaySubsContract;

import java.util.ArrayList;
import java.util.List;

public class SubItemAdapter extends RecyclerView.Adapter<SubItemAdapter.SubItemVH> {
  private List<SubredditModel> subList = new ArrayList<>();
  private DisplaySubsContract.SubAdapterDelegate subAdapterDelegate;

  public SubItemAdapter(DisplaySubsContract.SubAdapterDelegate subAdapterDelegate) {
    this.subAdapterDelegate = subAdapterDelegate;
  }

  class SubItemVH extends RecyclerView.ViewHolder {
    SubItemView subItemView;

    SubItemVH(SubItemView subItemView) {
      super(subItemView);
      this.subItemView = subItemView;

      subItemView.setOnClickListener(v ->
          subAdapterDelegate.interact(
              subList.get(getAdapterPosition()),
              DisplaySubsContract.SubInteraction.Visit.INSTANCE
          )
      );

      subItemView.setOnLongClickListener(v -> {
        subAdapterDelegate.interact(
            subList.get(getAdapterPosition()),
            DisplaySubsContract.SubInteraction.Preview.INSTANCE
        );
        return true;
      });
    }

    void bind(SubredditModel subModel) { subItemView.bind(subModel); }
  }


  @NonNull
  @Override
  public SubItemVH onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
    SubItemView subItemView = new SubItemView(parent.getContext());
    return new SubItemVH(subItemView);
  }

  @Override
  public void onBindViewHolder(@NonNull SubItemVH holder, int position) {
    holder.bind(subList.get(position));
  }

  @Override
  public int getItemCount() {
    return subList.size();
  }


  /**
   *
   * @param newSubs list of all posts
   */
  public void setList(List<SubredditModel> newSubs) {
      calculateDiff(newSubs).dispatchUpdatesTo(this);
      subList = newSubs;
  }

  private DiffUtil.DiffResult calculateDiff(List<SubredditModel>  newSubs) {
    return DiffUtil.calculateDiff(new DiffUtil.Callback() {
      @Override
      public int getOldListSize() {
        return subList.size();
      }

      @Override
      public int getNewListSize() {
        return newSubs.size();
      }

      @Override
      public boolean areItemsTheSame(int oldItemPosition, int newItemPosition) {
        return (subList.get(oldItemPosition).getId().equals(newSubs.get(newItemPosition).getId()));
      }

      @Override
      public boolean areContentsTheSame(int oldItemPosition, int newItemPosition) {
        return true;
      }
    });
  }

  @BindingAdapter({"bind:bannerUrl"})
  public static void loadImage(ImageView imgView, String bannerUrl) {
    // does not load image Wif the banner img string is empty
//    if (bannerUrl.length() > 0) {
//      try {
//        Log.d("SUB_ITEM_ADAPTER", "URL = " + bannerUrl);
//        Picasso.get().load(bannerUrl).into(imgView);
//      }
//      catch (Error e) {
//        Log.d("SUB_ITEM_ADAPTER", "Issue loading image " + e.toString());
//      }
//    }
  }

  public void clearList() {
    subList.clear();
    notifyDataSetChanged();
  }
}

File: app/src/main/java/com/relic/presentation/displaysubs/list/SubItemAdapter.kt
package com.relic.presentation.displaysubs.list

import android.content.Context
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.DiffUtil.DiffResult
import androidx.recyclerview.widget.RecyclerView
import com.relic.data.PostSource
import com.relic.domain.models.SubredditModel
import com.relic.interactor.Contract.SubAdapterDelegate
import com.relic.interactor.SubInteraction
import com.relic.presentation.base.RelicAdapter
import com.relic.presentation.displaysubs.pinned.SubsHeaderView
import kotlinx.coroutines.launch
import java.util.*

private const val TYPE_HEADER = 0
private const val TYPE_SUBS = 1

class SubItemAdapter(
    private val context: Context,
    private val subAdapterDelegate: SubAdapterDelegate
) : RelicAdapter<RecyclerView.ViewHolder>() {
    private val subsHeaderView: SubsHeaderView
    private var subList: MutableList<SubredditModel> = ArrayList()

    init {
        subsHeaderView = SubsHeaderView(context) { postSource ->
            subAdapterDelegate.interact(postSource, SubInteraction.Visit)
        }
    }

    override fun getItemViewType(position: Int): Int {
        return when (position) {
            0 -> TYPE_HEADER
            else -> TYPE_SUBS
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        return when (viewType) {
            TYPE_HEADER -> {
                SubsHeaderVH(subsHeaderView)
            }
            else -> {
                val subItemView = SubItemView(parent.context)
                SubItemVH(subItemView)
            }
        }
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        when (getItemViewType(position)) {
            TYPE_SUBS -> (holder as SubItemVH).bind(subList[position - 1])
        }
    }

    override fun getItemCount(): Int {
        // setting the item count to 0 when the sublist is empty prevents the recyclerview from
        // jumping to the very bottom once the actuals sub items are loaded
        return if (subList.isEmpty()) 0 else subList.size + 1
    }

    fun setList(newSubs: MutableList<SubredditModel>) {
        launch {
            calculateDiff(newSubs).dispatchUpdatesTo(this@SubItemAdapter)
            subList = newSubs
        }
    }

    fun setPinnedSubs(pinnedSubs: List<SubredditModel>) {
        subsHeaderView.setPinnedSubreddits(pinnedSubs)
    }

    private fun calculateDiff(newSubs: List<SubredditModel>): DiffResult {
        return DiffUtil.calculateDiff(object : DiffUtil.Callback() {
            override fun getOldListSize(): Int {
                return subList.size
            }

            override fun getNewListSize(): Int {
                return newSubs.size
            }

            override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
                return subList[oldItemPosition].id == newSubs[newItemPosition].id
            }

            override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
                return true
            }
        })
    }

    // region viewholders
    inner class SubItemVH(var subItemView: SubItemView) : RecyclerView.ViewHolder(subItemView) {
        fun bind(subModel: SubredditModel?) {
            subItemView.bind(subModel!!)
        }

        init {
            subItemView.setOnClickListener { v: View? ->
                subAdapterDelegate.interact(
                    PostSource.Subreddit(subList[absoluteAdapterPosition -1].subName),
                    SubInteraction.Visit)
            }
            subItemView.setOnLongClickListener { v: View? ->
                subAdapterDelegate.interact(
                    PostSource.Subreddit(subList[absoluteAdapterPosition - 1].subName),
                    SubInteraction.Preview
                )
                true
            }
        }
    }

    inner class SubsHeaderVH(
        subsHeaderView: SubsHeaderView
    ) : RecyclerView.ViewHolder(subsHeaderView)
    // endregion viewholders
}